### PR TITLE
Adds a new CHECK macro that abort()s on failure.

### PR DESCRIPTION
**Pull request by timothy-king**

### CODE REVIEW

File: test/unit/util/check_white.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+class CheckWhite : public CxxTest::TestSuite
+{
+ public:
+  const int kOne = 1;
+
+  // This test just checks that this statement compiles.
+  std::string TerminalCvc4Fatal() const
+  {
+    CVC4_FATAL() << "This test that CVC4 fatal can be terminal.";
```
Comment (aniemetz) [inline test/unit/util/check_white.h]: This sentence doesn't quite parse.

**Comment by timothy-king** [comment]
Experiment: https://godbolt.org/g/i3n71x

I manually confirmed that DCHECK as written should be compiled out entirely when optimizating when assertions are off.  Here is the somewhat simplified version:
```c++
#include <ostream>

#define CVC4_NO_RETURN [[noreturn]]
#define CVC4_PREDICT_FALSE(x) (__builtin_expect(x, false))

class FatalStream
{
 public:
  FatalStream(const char* function, const char* file, int line);
  CVC4_NO_RETURN ~FatalStream();

  std::ostream& stream();

 private:
  void Flush();
};

class OstreamVoider
{
 public:
  OstreamVoider() {}
  // The operator precedence between operator& and operator<< is critical here.
  void operator&(std::ostream&) {}
};

#define CVC4_FATAL_IF(cond, function, file, line) \
  CVC4_PREDICT_FALSE(!(cond))                     \
  ? (void)0 : OstreamVoider() & FatalStream(function, file, line).stream()

#define DCHECK(cond) \
  CVC4_FATAL_IF(false, __PRETTY_FUNCTION__, __FILE__, __LINE__)

int maybeCheckX(int x) {
    DCHECK(x > 5) << "x must be above 5";
    return x;
}
```
Running this with the options "-std=c++11 -O1" on gcc versions 4.8.1, 4.9.0, 5.1, 6.1 and 7.1, and clang versions 3.4.1, 3.6, 4.0, 5.0 yielded something like:
```c++
maybeCheckX(int):
  mov eax, edi
  ret
```

File: test/unit/util/check_white.h (hunk 1 above)
Comment (timothy-king) [reply test/unit/util/check_white.h]: Done.

**Comment by timothy-king** [comment]
PTAL

**Comment by mpreiner** [comment]
Looks good!

**Comment by aniemetz** [comment]
LGTM

**Pull request merged**